File: EduAI-LMS-Frontend/src/services/CourseService.js
import api from "../api/axiosConfig";

// =====================================================
// GET ALL COURSES
// =====================================================
// ADMIN USE
// =====================================================

const getAllCourses = () => {
    return api.get("/courses");
};


// =====================================================
// GET COURSES FOR STUDENT
// =====================================================
// STUDENT USE
//
// Returns only courses where the student is enrolled.
//
// Example:
// /api/courses/student/3
// =====================================================

const getStudentCourses = (studentId) => {
    return api.get(`/courses/student/${studentId}`);
};


// =====================================================
// GET COURSE BY ID
// =====================================================

const getCourseById = (id) => {
    return api.get(`/courses/${id}`);
};


// =====================================================
// CREATE COURSE
// =====================================================
// ADMIN USE
// =====================================================

const createCourse = (course) => {
    return api.post("/courses", course);
};


// =====================================================
// UPDATE COURSE
// =====================================================
// ADMIN USE
// =====================================================

const updateCourse = (id, course) => {
    return api.put(`/courses/${id}`, course);
};


// =====================================================
// DELETE COURSE
// =====================================================
// ADMIN USE
// =====================================================

const deleteCourse = (id) => {
    return api.delete(`/courses/${id}`);
};


// =====================================================
// EXPORT
// =====================================================

const CourseService = {

    getAllCourses,

    getStudentCourses,

    getCourseById,

    createCourse,

    updateCourse,

    deleteCourse
};

export default CourseService;

File: backend/src/main/java/backend/service/service/CourseService.java
package backend.service;

import backend.dto.CourseResponse;
import backend.entity.Course;
import backend.repository.CourseRepository;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CourseService {

    private final CourseRepository courseRepository;

    private final EnrollmentService enrollmentService;


    public CourseService(
            CourseRepository courseRepository,
            EnrollmentService enrollmentService
    ) {

        this.courseRepository =
                courseRepository;

        this.enrollmentService =
                enrollmentService;
    }


    // =====================================================
    // CREATE COURSE
    // =====================================================

    public Course createCourse(
            Course course
    ) {

        return courseRepository.save(course);
    }


    // =====================================================
    // GET ALL COURSES
    // =====================================================
    //
    // ADMIN USE
    //
    // Returns every course.
    //
    // =====================================================

    public List<CourseResponse> getAllCourses() {

        return courseRepository.findAll()
                .stream()
                .map(this::mapToResponse)
                .toList();
    }


    // =====================================================
    // GET COURSES FOR STUDENT
    // =====================================================
    //
    // ONLY COURSES WHERE THE STUDENT IS ENROLLED
    //
    // =====================================================

    public List<CourseResponse> getCoursesForStudent(
            Long studentId
    ) {

        List<Long> enrolledCourseIds =
                enrollmentService
                        .getStudentCourseIds(studentId);


        if (enrolledCourseIds.isEmpty()) {

            return List.of();
        }


        return courseRepository.findAll()
                .stream()
                .filter(course ->
                        course.getId() != null
                                &&
                        enrolledCourseIds.contains(
                                course.getId()
                        )
                )
                .map(this::mapToResponse)
                .toList();
    }


    // =====================================================
    // GET COURSE BY ID
    // =====================================================

    public Course getCourseById(
            Long id
    ) {

        return courseRepository.findById(id)
                .orElseThrow(
                        () -> new RuntimeException(
                                "Course not found"
                        )
                );
    }


    // =====================================================
    // CHECK STUDENT COURSE ACCESS
    // =====================================================

    public boolean studentCanAccessCourse(
            Long studentId,
            Long courseId
    ) {

        return enrollmentService
                .isStudentEnrolled(
                        studentId,
                        courseId
                );
    }


    // =====================================================
    // GET COURSE FOR STUDENT
    // =====================================================

    public Course getCourseForStudent(
            Long studentId,
            Long courseId
    ) {

        if (
                !studentCanAccessCourse(
                        studentId,
                        courseId
                )
        ) {

            throw new RuntimeException(
                    "Student is not enrolled in this course."
            );
        }


        return getCourseById(courseId);
    }


    // =====================================================
    // ENTITY -> DTO
    // =====================================================

    private CourseResponse mapToResponse(
            Course course
    ) {

        return new CourseResponse(

                course.getId(),

                course.getTitle(),

                course.getDescription(),

                course.getCourseCode(),

                course.getInstructor() != null
                        ? course.getInstructor().getId()
                        : null,

                course.getInstructor() != null
                        ? course.getInstructor().getFullName()
                        : null
        );
    }


    // =====================================================
    // UPDATE COURSE
    // =====================================================

    public Course updateCourse(
            Long id,
            Course updatedCourse
    ) {

        Course course =
                getCourseById(id);


        course.setTitle(
                updatedCourse.getTitle()
        );

        course.setDescription(
                updatedCourse.getDescription()
        );

        course.setInstructor(
                updatedCourse.getInstructor()
        );


        return courseRepository.save(course);
    }


    // =====================================================
    // DELETE COURSE
    // =====================================================

    public void deleteCourse(
            Long id
    ) {

        Course course =
                getCourseById(id);

        courseRepository.delete(course);
    }
}